### PR TITLE
Fetch api

### PR DESCRIPTION
**Pull request by jql6**
# Summary
This pull request adds a component that fetches a JSON file from a [Mocky API](https://run.mocky.io/v3/02e9fb29-4516-48b1-8bd2-43e868303cb6) and then displays it in a table.

Fetch_API

# Table of contents
- [Mocky API](#mocky-api)
- [`FetchComponent.js`](#fetchcomponentjs)
- [Lock files](#lock-files)
- [Usage](#usage)

# Mocky API
The JSON file is pulled from a mock API from [Mocky](https://run.mocky.io/v3/02e9fb29-4516-48b1-8bd2-43e868303cb6).

JSON_file

# `FetchComponent.js`
This file contains a component comprised of a button and a table. The button gets the JSON file from the [Mocky API](https://run.mocky.io/v3/02e9fb29-4516-48b1-8bd2-43e868303cb6) and the table displays the JSON data after the JSON is fetched.

# Lock files
The `yarn.lock` and `package-lock.json` files are added for reproducibility.

# Usage
The app can be started from a fresh git clone by doing:
```
npm install
npm start
```
`npm install` is required if your folder doesn't have the `node_modules` folder yet.

### CODE REVIEW

File: src/components/FetchComponent.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+          login: null,
+        },
+        permissions: {
+          roles: null,
+        },
+      },
+    };
+  }
+  /* Fetches json from mocky API */
+  async fetchJSON() {
```
Comment (reneriosp) [inline src/components/FetchComponent.js]: This needs to be in a different file and needs to take parameters, improve error handling.
Also take advantage of the Promise's status, for loaders, etc.

File: src/App.js [hunk 2]
```diff
@@ -1,25 +1,16 @@
...
-        </a>
-      </header>
-    </div>
-  );
+class App extends Component {
+
+  render() {
+    return (
+      <div>
+        <FetchComponent title="Fetch API" />
```
Comment (reneriosp) [inline src/App.js]: Improve Component naming, the component will be representing the table, not the fetch process.

File: src/components/FetchComponent.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    /* For debugging */
+    console.log(this.state.data);
+  }
+
+  render() {
+    return (
+      <div>
+        <h1>{this.props.title}</h1>
+        <button onClick={this.fetchJSON}>Get data</button>
+        <table border={2} cellPadding={5}>
```
Comment (reneriosp) [inline src/components/FetchComponent.js]: Not showing the user that data is being loaded.
As data will come asynchronously you need to show the user that you are waiting for data.

**Review by reneriosp — COMMENTED** [review]
Quick review

File: src/components/FetchComponent.js (hunk 1 above)
Comment (jql6) [reply src/components/FetchComponent.js]: I've factored `fetchJSON()` out of the `FetchComponent.js` file into a new file called [`fetchJSON.js`](https://github.com/jql6/mentorsNetwork-FrontEnd/blob/FetchAPI/src/functions/fetchJSON.js) in the `functions` folder.

fetchJSON

It now takes a parameter for the URL and uses the response status. It also throws an error if the response isn't 200 (OK).

File: src/App.js (hunk 2 above)
Comment (jql6) [reply src/App.js]: The `FetchComponent` component has been renamed to [`UserTable`](https://github.com/jql6/mentorsNetwork-FrontEnd/blob/FetchAPI/src/components/UserTable.js) to better represent its purpose.

File: src/components/FetchComponent.js (hunk 3 above)
Comment (jql6) [reply src/components/FetchComponent.js]: I've added a loading indicator while the function is fetching the JSON data in the renamed [`UserTable`](https://github.com/jql6/mentorsNetwork-FrontEnd/blob/FetchAPI/src/components/UserTable.js) component file.

This is done through a `loading` status which is set to `true` in `loadData()` before the json file is obtained, and set back to `false` once the file is obtained.
Loading

When the `loading` status is `true`, the element displays "Loading...".
loading 2

Here's a demo of the loading indicator in the app:
Loading demo

File: src/components/UserTable.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  /* Placeholder table until we load the data */
+  const [data, setData] = useState({ users: [makeUser()] });
+  const [loading, setLoading] = useState(false);
+
+  async function loadData() {
+    setLoading(true);
+    // Downloading json
+    try {
+      const JSONData = await getJSON();
+      await new Promise((resolve) => {
```
Comment (reneriosp) [inline src/components/UserTable.js]: The promise needs to be returned by the function, not later.

File: src/functions/getJSON.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+// getJSON.js
+async function getJSON(url = "http://localhost:3001/db") {
```
Comment (reneriosp) [inline src/functions/getJSON.js]: I would prefer to consolidate functions in a class, not 2 different files

**Review by reneriosp — CHANGES_REQUESTED** [review]

**Pull request closed without merging**